LFS171x/fabric-material/chaincode/tuna-app/tuna-chaincode.go

```
@@ -0,0 +1,190 @@
+
+	if len(args) != 5 {
+		return shim.Error("Incorrect number of arguments. Expecting 5")
+	}
+
+	var tuna = Tuna{ Vessel: args[1], Location: args[2], Timestamp: args[3], Holder: args[4] }
+
+	tunaAsBytes, _ := json.Marshal(tuna)
+	APIstub.PutState(args[0], tunaAsBytes)
+
+	return shim.Success(nil)
+}
+
+func (s *SmartContract) queryAllTuna(APIstub shim.ChaincodeStubInterface) sc.Response {
+
+	startKey := "0"
+	endKey := "999"
+
+	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+	defer resultsIterator.Close()
+
+	// buffer is a JSON array containing QueryResults
+	var buffer bytes.Buffer
+	buffer.WriteString("[")
+
+	bArrayMemberAlreadyWritten := false
+	for resultsIterator.HasNext() {
+		queryResponse, err := resultsIterator.Next()
+		if err != nil {
+			return shim.Error(err.Error())
+		}
+		// Add a comma before array members, suppress it for the first array member
+		if bArrayMemberAlreadyWritten == true {
+			buffer.WriteString(",")
+		}
+		buffer.WriteString("{\"Key\":")
+		buffer.WriteString("\"")
+		buffer.WriteString(queryResponse.Key)
+		buffer.WriteString("\"")
+
+		buffer.WriteString(", \"Record\":")
+		// Record is a JSON object, so we write as-is
+		buffer.WriteString(string(queryResponse.Value))
+		buffer.WriteString("}")
+		bArrayMemberAlreadyWritten = true
+	}
+	buffer.WriteString("]")
+
+	fmt.Printf("- queryAllTuna:\n%s\n", buffer.String())
+
+	return shim.Success(buffer.Bytes())
+}
+
+func (s *SmartContract) changeTunaHolder(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
+
+	if len(args) != 2 {
+		return shim.Error("Incorrect number of arguments. Expecting 2")
+	}
+
+	tunaAsBytes, _ := APIstub.GetState(args[0])
+	tuna := Tuna{}
+
+	json.Unmarshal(tunaAsBytes, &tuna)
+	tuna.Holder = args[1]
+
+	tunaAsBytes, _ = json.Marshal(tuna)
+	APIstub.PutState(args[0], tunaAsBytes)
+
+	return shim.Success(nil)
```

We should check the results of the call to PutState and only return success if the PutState call was successful; otherwise, we should return an error.

---

I added this in:

``` 	err := APIstub.PutState(args[0], tunaAsBytes)

    if err != nil {
            return "", fmt.Errorf("Failed to set asset: %s", args[0])
    } ```